### ISSUE TITLE
Error. AZSK module is not available in automation account

### ISSUE THREAD

**Issue by akamalov** [azsk/DevOpsKit#718]
## Description</br>

Trying to install AzSK Continuous Assurance Automation account:

## Steps to reproduce

Create AzSK Continuous Assurance Account:

```
PS C:\Users\azureuser> Install-AzSKContinuousAssurance -SubscriptionId $subId -AutomationAccountLocation $location -AutomationAccountRGName $rgname -ResourceGroupNames 'AK-TEST-0001,captain-america,aaas-rg' -OMSWorkspaceId $lawsId -OMSSharedKey $omsKey -AzureADAppName 'AzSk-Assurance' -ScanIntervalInHours 24

Auto-update for AzSK is currently not enabled for your machine. To set it, run the command below:
Set-AzSKPolicySettings -AutoUpdate On

A newer version of AzSK is available: Version 4.0.0
To update, run the command below in a fresh PS window:
Install-Module -Name AzSK -Scope CurrentUser -AllowClobber -Force
Using the latest version ensures that AzSK security commands you run use the latest, most up-to-date controls.
Results from the current version should not be considered towards compliance requirements.
================================================================================
================================================================================
AzSK Version: 3.15.0
================================================================================
Method Name: Install-AzSKContinuousAssurance (ICA)
Input Parameters:
Name                      Alias Value
----                      ----- -----
SubscriptionId            sid   XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX
AutomationAccountLocation loc   eastus2
AutomationAccountRGName   aargn AK-AzSK-Test-001
ResourceGroupNames        rgns  AK-TEST-0001,aaas-rg
LAWSId                    wid   XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXXXX
LAWSSharedKey             wkey  XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
AzureADAppName            spn   AzSk-Assurance
ScanIntervalInHours       si    24

You can also use: ica -sid XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX -loc eastus2 -aargn AK-AzSK-Test-001 -rgns AK-TEST-0001,aaas-rg -wid XXXXXXXX-XXXX-X
 -si 24
================================================================================
Running AzSK cmdlet using a generic (org-neutral) policy...
================================================================================
Started setting up Automation Account for Continuous Assurance (CA)
================================================================================
Creating Automation Account: [AzSKContinuousAssurance]
Found AAD application in the directory: [AzSk-Assurance]
Generating new credential for AzSK CA SPN
Configuring permissions for AzSK CA SPN. This may take a few min...
Adding SPN to [Contributor] role at [AzSKRG] resource group scope...
WARNING: Ignoring error while assigning CA SPN permissions for SPN: [XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX].
WARNING: Make sure this SPN is 'Contributor' on AzSKRG and 'Reader' on the subscription.
Successfully configured AzSK CA Automation Account with SPN.
Creating a storage account: [azsk201908XXXXXXXX] for storing reports from CA scans.
Updating runbook: [Continuous_Assurance_Runbook]
--------------------------------------------------------------------------------
Completed setup phase-1 for AzSK Continuous Assurance.
Setup phase-2 has been triggered and will continue automatically in the background. This involves loading all PS modules CA requires to run, scheduling runbook, etc
You can check the overall status of installation using the 'Get-AzSKContinuousAssurance' command 2 hours after running 'Install-AzSKContinuousAssurance' command.
Once phase-2 setup completes, your subscription and resources (from the specified resource groups) will be scanned periodically by CA. All security control evaluati
You may subsequently update any of the parameters specified during installation using the 'Update-AzSKContinuousAssurance' command. If you specified '*' for resourc
You should use the AzSK Monitoring solution to monitor your subscription and resource health status.
================================================================================
Logs have been exported to: 'C:\Users\azureuser\AppData\Local\Microsoft\AzSKLogs\Sub_Enterprise\20190814_130552_ICA'
================================================================================
C:\Users\azureuser\AppData\Local\Microsoft\AzSKLogs\Sub_Enterprise\20190814_130552_ICA
PS C:\Users\azureuser>

```

Wait 2hours for phase2 to get completed. Display AzSK Continuous Assurance Account:

```
PS C:\Users\azureuser> Get-AzSKContinuousAssurance -AutomationAccountName AzSKContinuousAssurance -SubscriptionId $subId -AutomationAccountRGName $rgname

Auto-update for AzSK is currently not enabled for your machine. To set it, run the command below:
Set-AzSKPolicySettings -AutoUpdate On

A newer version of AzSK is available: Version 4.0.0
To update, run the command below in a fresh PS window:
Install-Module -Name AzSK -Scope CurrentUser -AllowClobber -Force
Using the latest version ensures that AzSK security commands you run use the latest, most up-to-date controls.
Results from the current version should not be considered towards compliance requirements.
================================================================================
================================================================================
AzSK Version: 3.15.0
================================================================================
Method Name: Get-AzSKContinuousAssurance (GCA)
Input Parameters:
Name                    Alias Value
----                    ----- -----
AutomationAccountName   aan   AzSKContinuousAssurance
SubscriptionId          s     XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX
AutomationAccountRGName aargn AK-AzSK-Test-001

You can also use: gca -aan AzSKContinuousAssurance -s XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX -aargn AK-AzSK-Test-001
================================================================================
Running AzSK cmdlet using a generic (org-neutral) policy...
================================================================================
Started validating your AzSK Continuous Assurance (CA) setup...
================================================================================
Check 01: Presence of CA Automation Account.
Status:   OK. Found the CA Automation Account: [AzSKContinuousAssurance].
--------------------------------------------------------------------------------
Check 02: Checking CA Runbook version.
Status:   OK. CA runbook is healthy.
--------------------------------------------------------------------------------
Check 03: Inspecting CA module: [AZSK].
Status:   Failed. AZSK module is not available in automation account.
To resolve this please run command 'Remove-AzSKContinuousAssurance' followed by 'Install-AzSKContinuousAssurance'.
--------------------------------------------------------------------------------
Summary of CA configuration:
Name                          Value
----                          -----
AltLAWSId                     NULL
AppResourceGroupNames         AK-TEST-0001,aaas-rg
AutomationAccountName         AzSKContinuousAssurance
AzSKReportsStorageAccountName azsk201908XXXXXXXX
AzureADAppID                  XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX
AzureADAppName                AzSk-Assurance
CertificateExpiry             2/15/2020 1:05:59 PM -05:00
LAWSId                        XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX
Runbooks                      Continuous_Assurance_Runbook
RunbookVersion                Current version: [3.1902.0] Latest version: [3.1902.0]
Schedules                     CA_Scan_Schedule (Frequency: 24 Hour)
WebhookUrl                    NULL
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
Found that AzSK Continuous Assurance (CA) is not correctly setup or functioning properly.
Review the failed check and follow the remedy suggested. If it does not work, please file a support request after reviewing the FAQ.
--------------------------------------------------------------------------------
================================================================================
Logs have been exported to: 'C:\Users\azureuser\AppData\Local\Microsoft\AzSKLogs\Sub_Enterprise\20190814_151456_GCA'
================================================================================
C:\Users\azureuser\AppData\Local\Microsoft\AzSKLogs\Sub_Enterprise\20190814_151456_GCA
PS C:\Users\azureuser>

```

As you can see the account creation failed. I don't know what is going on here. Why would it fail?

:(

**Comment by akamalov** [comment]
I went ahead, deleted and re-run ```Install-AzSKContinuousAssurance```, but I got the same thing all over again:

```
Status:   Failed. AZSK module is not available in automation account.
To resolve this please run command 'Remove-AzSKContinuousAssurance' followed by 'Install-AzSKContinuousAssurance'
```
I do not know what is causing a setup to fail. There is no detailed debug output.

**Comment by akamalov** [comment]
Loaded all AzSK dependency modules manually, under my automation account. Still cannot import ```AzSK``` module because it is complaining that ``` This module has dependencies that are not present in this account. All dependencies must be present before this module can be imported.```

**Example:**

```
This module has dependencies that are not present in this account. All dependencies must be present before this module can be imported.

Dependencies:

Az.Accounts (= 1.6.0)
Az.Storage (= 1.5.0)
Az.AnalysisServices (= 1.1.0)
Az.ApiManagement (= 1.2.0)
Az.ApplicationInsights (= 1.0.0)
Az.Automation (= 1.3.0)
Az.Batch (= 1.1.0)
Az.Cdn (= 1.3.0)
Az.Compute (= 2.4.0)
Az.DataFactory (= 1.1.2)
Az.DataLakeAnalytics (= 1.0.0)
Az.DataLakeStore (= 1.2.1)
Az.EventHub (= 1.2.0)
Az.HDInsight (= 2.0.0)
Az.Monitor (= 1.2.1)
Az.KeyVault (= 1.2.0)
Az.LogicApp (= 1.2.1)
Az.Network (= 1.11.0)
Az.NotificationHubs (= 1.1.0)
Az.OperationalInsights (= 1.3.1)
Az.PolicyInsights (= 1.1.2)
Az.RedisCache (= 1.1.0)
Az.Resources (= 1.6.0)
Az.ServiceBus (= 1.2.1)
Az.ServiceFabric (= 1.1.1)
Az.Sql (= 1.13.0)
Az.StreamAnalytics (= 1.0.0)
Az.TrafficManager (= 1.0.1)
Az.Websites (= 1.3.0)
Az.ContainerInstance (= 1.0.1)
Az.ContainerRegistry (= 1.0.1)
```

However, I did check and I do have all dependency modules ```Available``` under this automation account. However, I cannot import ```AzSK``` module, as it is greyed out...

**Comment by akamalov** [comment]
I was able to import AzSK from powershell gallery and uploading to automation:

![image](https://user-images.githubusercontent.com/7885257/63168672-c63f1000-c002-11e9-9e05-a2e43399f62e.png)

Importing AzSK from Azure gallery has failed repeatedly, even though dependencies were satisfied.

**Comment by ganesh-msft** [comment]
Can you please go to  Continuous_Assurance_Runbook in AzSKRG and see if any error is reported there. Usually it installs AzSK module after completing 4-5 iterations of the runbook jobs. If no error and you don't see as many iterations, you can manually trigger them multiple times using the 'play' icon at the toprather than waiting for 2 hrs.  If it still doesn't succeed, please share the errors that you see in the runbook log.

**Comment by akamalov** [comment]
Thanks @ganesh-msft for your prompt comment. I solved the problem using the following:

1. I had to make sure that my account had ```owner``` on subscription. This however is a bit difficult to achieve because we have different subscriptions owned by different teams and having ```owner``` on subscription is very hard bargain that one has to pull. Having ```owner``` on subscription is under very ideal situations which is far from realities of production environment.

2. I had to manually push AzSK using https://www.powershellgallery.com/packages/AzSK/3.15.0 and selecting ```Deploy to Azure Automation```. Selecting AzSK off Azure gallery and selecting ```import``` never worked due to module dependencies.

Once these two obstacles overcame, AzSK Continuous Assurance was installed without a problem.

Thanks a lot.

Alex
